### PR TITLE
[ML] Add indices_options to datafeed config and update

### LINKED ISSUE #48056: [ML] Allow `indices_options` parameter for datafeed configuration
In certain situations it is valuable to allow missing indices (e.g. a well defined datafeed configuration that takes advantage of ECS). Right now, if one of the indices in the supplied index pattern does not exist, we throw an error on `_start`. If we exposed an `indices_options` type of parameter (that serializes to an `IndicesOptions` object), that will allow more advanced users to provide custom expansion options to handle their use case.

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new configurable field called `indices_options`. This allows users to create or update the indices_options used when a datafeed reads from an index. 

This is necessary for the following use cases:
 - Reading from frozen indices
 - Allowing certain indices in multiple index patterns to not exist yet

These index options are available on datafeed creation and update. Users may specify them as URL parameters or within the configuration object.
 
closes https://github.com/elastic/elasticsearch/issues/48056

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPutDatafeedAction.java [hunk 1]
```diff
@@ -44,8 +46,15 @@ public String getName() {
     @Override
     protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
         String datafeedId = restRequest.param(DatafeedConfig.ID.getPreferredName());
+        IndicesOptions indicesOptions = null;
+        if (restRequest.hasParam("expand_wildcards")
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPutDatafeedAction.java]: Could we avoid checking if `restRequest` has those params? I think it'd work to do `IndicesOptions indicesOptions = IndicesOptions.fromRequest(restRequest, SearchRequest.DEFAULT_INDICES_OPTIONS)`

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestUpdateDatafeedAction.java [hunk 2]
```diff
@@ -44,8 +46,15 @@ public String getName() {
     @Override
     protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
         String datafeedId = restRequest.param(DatafeedConfig.ID.getPreferredName());
+        IndicesOptions indicesOptions = null;
+        if (restRequest.hasParam("expand_wildcards")
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestUpdateDatafeedAction.java]: ditto

**Review by dimitris-athanasiou — COMMENTED** [review]
Looks good! My only confusion has to do with handling default options when it is not set. In particular, it seems we used to have 2 different defaults in different places: IndicesOptions.lenientExpandOpen() and SearchRequest.DEFAULT_INDICES_OPTIONS, which is IndicesOptions.strictExpandOpenAndForbidClosedIgnoreThrottled().

With this change defaults will remain the same when there are no options set. But if a user sets options explicitly, we use SearchRequest.DEFAULT_INDICES_OPTIONS as the default options. This is the part that confuses me.

**Comment by droberts195** [comment]
> it seems we used to have 2 different defaults in different places: IndicesOptions.lenientExpandOpen() and SearchRequest.DEFAULT_INDICES_OPTIONS, which is IndicesOptions.strictExpandOpenAndForbidClosedIgnoreThrottled().

It seems that this was a bug before, and it didn't really matter in the days before frozen indices but frozen indices mean it could cause weird behaviour.

We discussed this a bit in a different channel and the outcome was:

1. We _should_ use the same default in both places, and that default should be `SearchRequest.DEFAULT_INDICES_OPTIONS`
2. As a result of this, the `indicesOptions` member of `DatafeedConfig` can be made `@NotNull` and defaulted to `SearchRequest.DEFAULT_INDICES_OPTIONS` at an earlier stage - this will reduce the number of `null` checks required downstream
3. We have an extra check for no concrete indices in addition to the indices options check: https://github.com/elastic/elasticsearch/blob/7e4c0e699171d8dd93b0bf0d2e6939231cd67683/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedNodeSelector.java#L125-L127

Currently that sets the same assignment failure reason as an exception from the indices options check sets, and this could lead to confusion if indices options include `allow_no_indices`.  So the reason returned on line 126 of current master/line 128 of commit 1 on this PR should be made different to the reason returned on line 130 of current master/line 132 of commit 1 on this PR.  The current reason is most suited to the `if (concreteIndices.length == 0)` check and the new reason for the resolver throwing should include the `toString()` representation of the caught exception so that it reflects the problem given the configured indices options.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestUpdateDatafeedAction.java (hunk 2 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestUpdateDatafeedAction.java]: I don't agree. If none are provided, then that means the user doesn't want to update them.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPutDatafeedAction.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPutDatafeedAction.java]: Makes sense. DatafeedConfig can make the options non-null and default to `SearchRequestt.DEFAULT_INDICES_OPTIONS`

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestUpdateDatafeedAction.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestUpdateDatafeedAction.java]: Ah, yes. That's the update. My bad!

**Review by droberts195 — APPROVED** [review]
LGTM

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM2

**Pull request merged**